File: source/Flowlayoutdemo/app/src/main/java/com/zero/flowlayoutdemo/MainActivity.java
package com.zero.flowlayoutdemo;

import android.content.Intent;
import android.content.pm.PackageManager;
import android.os.Bundle;
import android.os.Trace;
import android.util.Log;
import android.view.View;
import android.widget.Button;

import androidx.appcompat.app.AppCompatActivity;
import androidx.core.app.ActivityCompat;

public class MainActivity extends AppCompatActivity {

    private static final int REQUEST_EXTERNAL_STORAGE = 1;
    private static String[] PERMISSIONS_STORAGE = {
            "android.permission.READ_EXTERNAL_STORAGE",
            "android.permission.WRITE_EXTERNAL_STORAGE"};

    public static void verifyStoragePermissions(AppCompatActivity activity) {


        try {
            //检测是否有写的权限
            int permission = ActivityCompat.checkSelfPermission(activity,
                    "android.permission.WRITE_EXTERNAL_STORAGE");
            if (permission != PackageManager.PERMISSION_GRANTED) {
                // 没有写的权限，去申请写的权限，会弹出对话框
                ActivityCompat.requestPermissions(activity, PERMISSIONS_STORAGE, REQUEST_EXTERNAL_STORAGE);
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    private Button mButton;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        verifyStoragePermissions(this);
        super.onCreate(savedInstanceState);
        Log.i(TAG, "onCreate: "+ this.getClass().getSimpleName());
        setContentView(R.layout.activity_main);
        mButton = findViewById(R.id.btn_list);
        mButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
//                android.os.Debug.stopMethodTracing();
                Intent intent = new Intent(MainActivity.this,RecyclerActivity.class);
                startActivity(intent);
            }
        });
        Trace.endSection();
    }

    @Override
    protected void onResume() {
        super.onResume();
        Log.i(TAG, "onResume: " + this.getClass().getSimpleName());
    }

    @Override
    protected void onStop() {
        super.onStop();
        Log.i(TAG, "onStop: "+ this.getClass().getSimpleName());
    }

    @Override
    protected void onPause() {
        super.onPause();
        Log.i(TAG, "onPause: "+ this.getClass().getSimpleName());
    }
    private static final String TAG = "Zero";

    @Override
    protected void onDestroy() {
        super.onDestroy();
        Log.i(TAG, "onDestroy: "+ this.getClass().getSimpleName());
    }
}

File: source/bleDemo/app/src/main/java/com/example/bledemo/MainActivity.kt
package com.example.bledemo

import android.Manifest
import android.bluetooth.BluetoothAdapter
import android.bluetooth.BluetoothDevice
import android.bluetooth.BluetoothManager
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.os.Build
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.util.Log
import android.view.View
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat

class MainActivity : AppCompatActivity() {

    companion object{
        const val TAG = "Zero"
    }

    private fun requestPermission(){
        if(Build.VERSION.SDK_INT >= Build.VERSION_CODES.M){
            if(ContextCompat.checkSelfPermission(this,
                Manifest.permission.ACCESS_FINE_LOCATION )!= PackageManager.PERMISSION_GRANTED){
                toast("自Android6开始需要打开位置权限才可以搜索到Ble设备")
            }
            ActivityCompat.requestPermissions(this,
            arrayOf(Manifest.permission.ACCESS_FINE_LOCATION,
            Manifest.permission.ACCESS_COARSE_LOCATION),666)
        }
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        if(requestCode == REQUEST_CODE_LOCATION_SETTING){
            if(isLocationEnable()){
                Log.i(TAG,"定位已经打开了")
            }else{
                Log.i(TAG,"定位没有打开")
            }
        }else super.onActivityResult(requestCode, resultCode, data)
    }

    override fun onRequestPermissionsResult(
        requestCode: Int,
        permissions: Array<out String>,
        grantResults: IntArray
    ) {
        if(requestCode == 666){
            if(grantResults.isNotEmpty() && grantResults[0] ==PackageManager.PERMISSION_GRANTED){
                toast("授权成功")
            }else{
                toast("授权失败")
                setLocationService()
            }
        }else{
            super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        }
    }


    private val  deviceList = mutableListOf<BluetoothDevice>()

    private val bluetoothAdapter by lazy {
        val bluetoothManager = getSystemService(Context.BLUETOOTH_SERVICE) as BluetoothManager
        Log.i("Zero","bluetoothManager: $bluetoothManager")
        bluetoothManager.adapter
    }


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        requestPermission()
    }

    fun  onScan(view: View): Unit{

        Log.i("Zero","onScan")
        deviceList.clear()

//        bluetoothAdapter.startLeScan(scanCallback)

        showBle.show(supportFragmentManager,"ddd")
    }

    private val showBle = RecyclerDialogFragment()
    private val scanCallback = BluetoothAdapter.LeScanCallback { device, rssi, scanRecord ->
        Log.i(TAG,"name: ${device.name} , mac: ${device.address}")
        if(!deviceList.contains(device)){
            runOnUiThread {
                deviceList.add(device)
                showBle.datas = deviceList
            }
        }
    }

    fun onStopScan(view: View): Unit{
        bluetoothAdapter.stopLeScan { device, rssi, scanRecord ->

        }
        showBle.dismiss()
    }


}
